### ISSUE TITLE
Make the housenumber quest faster

### ISSUE THREAD

**Issue by IpswichMapper** [westnordost/StreetComplete#2131]
The keyboard does not show up when you click on the task. 

This is particularly a problem for the housenumber quest, when multiple houses are very close together on residential streets as well as being on both sides. Due to the density of houses, you should be able to enter the numbers as quickly as possible so that you can complete lots of housenumber quests.

This may not be problem for other quests, which aren't so densely packed, but this one addition to the housenumber quest at least would be really useful, it could mean I can walk less slowly when surveying. (And get more houses done in the same amount of time).

**Comment by westnordost** [comment]
Well, the keyboard does not show up immediately for the same reason why large forms to input the quest answer (for example roof shape, sport, surface, ...) do not immediately take up all the screen space but first only 1/3 or so and you have to pull it up or tap on it to bring it to full-screen:

1. Users must be able to see the geometry of the element they are asked for on the map. Very important for the surface quest for example. Also, important for being able to orient oneself in relation to the element on the map
2. especially new users don't know what icon stands for which question, so they may just want to read the question first and then exit again. If reading the question immediately results in a new screen being shown, this would be very disruptive.

So I don't know how to accomodate these things with the desire to make inputting housenumbers a little more efficient.

I hear Android 11 has a new feature that the keyboard can come up smoothly (maybe it is necessary to implement something for that) instead of creating a visible yank. Maybe that would help?

**Comment by IpswichMapper** [comment]
I played around with streetcomplete and I understand the first problem.

For point 2, the solution is the make it an opt-in setting. This way new users aren't confused, whereas better users can turn the option on.

Can you make automatic keyboard an option in settings?

**Comment by westnordost** [comment]
Maybe we can find another solution than adding a setting, because this is not really a preference, point 1 is still important. Perhaps something like that the form remembers if you had the keyboard open the last time you used it and "restores" the keyboard if it was?

**Comment by peternewman** [comment]
If you wanted to really optimise the house number quest, I don't know how generic it is, but in the UK you usually either have odd numbers one side and evens the other or just sequential. What about (like the building heights), listing presets at the bottom like so:

So if I entered 42 last time:
40|41|43|44
---|---|---|---

Clearly it won't handle 42a, or 44-48 etc, but for single houses and semi-detached a lot of the time it could make it far quicker and more accurate than typing. It could even potentially learn your history to offer just the one direction.

I guess it could always be disabled for places that will have more than one number.

Obviously like the building height quest, you could tap in the box to get the keyboard up.

**Comment by westnordost** [comment]
Well that's an ok idea, but there is not really space for 4 reasonable spaced buttons. There is already the ABC-button.

I think the main problem with the keyboard is that it is always so disruptive when it opens up. But that can maybe be solved with Android 11.

**Comment by IpswichMapper** [comment]
@westnordost 
What do you mean to say that "the form remembers you had the keyboard open"? I don't really understand. Do you mean to say that if you have done one housenumber quest, the keyboard will automatically open for the next houseumber quest?

What is this android 11 thing you are talking about? Can you link me to a demonstration? (Furthermore, most people don't have android 11 , there is also iphone to worry about.
<hr
About the buttons proposal: 

There are two reasons I think streetcomplete is so good for house number mapping.

1. You can verify which building has which housenumber on site, so that you are not left with a mess of housenumbers (e.g. when using software such as keypad mapper). 
2. Its low barrier of entry and cross platform support. This, I think, is really, really valuable. Housenumbers are everywhere to be found. That means that geography limits how many housenumbers one can survey. However, drawing buildings is something anyone can do using the imagery. This means that better mappers can draw the buildings, while less experienced, more local mappers can do the surveying.

But for Streetcomplete to be practical at this job, it needs to be as efficient. Currently, I can map around 3-3.5km of housenumbers in an hour. That's not a lot (on average many people walk twice that fast).

@peternewman 

This proposal seems good. If it were to be implemented, the textbox could be removed, leaving only numbers (with the "abc" button replaced by a keyboard button).

One problem, however, is that housenumbers on one side of the road are very different to the other side.

For example, I could have 52, 54, 56 on one side and 41, 43, 45 on the other.

Instead, if this were to be implemented, then odd and even numbers would need to be remembered seperately.

For example, if I had entered 52 and 41, the options would be (39, 43, 50, 56).

This becomes significantly more complicated. Also, what about street changes? Would this system work in other countries

This system would also not work on streets where housenumbers 


Looking at the housenumbers in New York, they seem to follow the same odds on one side, evens on the other.

However, because of the right angled roads, roads can continue for ages ,meaning housenumbers reach in the 1000s. This is clearly infeasable to enter one by one, so some system has to be devised to solve this.

On top of this, extra buttons can be confusing. Maybe they should not be enabled by default, amd left as a option, so more accustomed users to Streetcomplete can turn it on

**Comment by matkoniecz** [comment]
> What is this android 11 thing you are talking about? Can you link me to a demonstration? (Furthermore, most people don't have android 11)

Probably "Android 11: Smooth keyboard appear animation enhancement" at #2133

**Comment by IpswichMapper** [comment]
That kind of makes sense, thanks.

**Comment by peternewman** [comment]
> Well that's an ok idea, but there is not really space for 4 reasonable spaced buttons.

The keyboard has 10 across the top row? Admittedly showing less characters. Or lay them out as:

-2|-1|+1|+2
--|--|--|--

Or

40|41
---|---
**43**|**44**

Compared to the screen real estate taken up by the keyboard itself there is a **lot** of space available!

> There is already the ABC-button.

I'm not sure if that was aimed at me, but on that front, shouldn't 123 be the default for house**number**?

> But for Streetcomplete to be practical at this job, it needs to be as efficient. Currently, I can map around 3-3.5km of housenumbers in an hour. That's not a lot (on average many people walk twice that fast).

Agreed entirely, I know SC generally feels fairly speedy, but recently I was in an area where someone had done the outlines and numbers, but none of the detail, it was loads faster than anything else I can imagine, but interestingly I could still walk quicker than I could even enter that it was a house, let alone tagging building heights (which #1772 will help address).

> This proposal seems good. If it were to be implemented, the textbox could be removed, leaving only numbers (with the "abc" button replaced by a keyboard button, which shows the keyboard and the texbox instead of buttons).

Yeah I was imagining something like that. You need to be able to enter letters too, although I guess it could offer a-f say as buttons to append a letter onto a house number too (a bit like a Stenographer but simpler), although I guess it only needs the first or last letter probably.

> One problem, however, is that housenumbers on one side of the road are very different to the other side.

Agreed, although do you map both sides at once? With the building heights I'd glance across sometimes, but I think you might need to cross to read a small or obscured number, although I guess people would like to avoid mapping each road twice if possible.

> Instead, if this were to be implemented, then odd and even numbers would need to be remembered seperately.

Yeah, perhaps a toggle to the other side of the road (i.e. sequence B option) if people do number both sides of the road at the same time.

> For example, if I had entered 52 and 41, the options would be (39, 43, 50, 54).

Sorry, I was actually suggesting it just offers n-2, n-1, n+1 and n+2 on your previous number. So if you last entered 41 you'd have 39, 40, 42, 43. But yes, if you're doing both sides simultaneously, some heuristics or way to do something clever would be good, if it actually happens.

> This becomes significantly more complicated. Also, what about street changes?

When the street changes, you go to the keyboard and start over. I only imagined it as a way to avoid the keyboard 90% of the time, which will provide a huge speed-up.

> Would this system work in other countries

You can still use the keyboard! It would work anywhere that has house numbers which use latin numbers and increment sequentially.

> This system would also not work on streets where housenumbers
...?

> Looking at the housenumbers in New York, they seem to follow the same odds on one side, evens on the other.

There are places in the UK where they are just sequential, so 1, 2, 3 but that works fine too.

Looks like it wouldn't work in Russia:
https://wiki.openstreetmap.org/wiki/Addresses#Letters_in_house_numbers_.28for_example_in_Russia.29

> However, because of the right angled roads, roads can continue for ages ,meaning housenumbers reach in the 1000s. This is clearly infeasable to enter one by one, so some system has to be devised to solve this.

There is interpolation:
https://wiki.openstreetmap.org/wiki/Addresses#Using_interpolation

> On top of this, extra buttons can be confusing. Maybe they should not be enabled by default, amd left as a option, so more accustomed users to Streetcomplete can turn it on.

Possibly, we'd have the same issue with #1772 . If we keep the text box, it's pretty clear what the buttons do as soon as you've pressed one! I think the normal behaviour is two step for stuff like this, so it would be a short-cut to pre-fill the box, then you'd submit as normal. Which would still be faster than typing for me, and probably more accurate most of the time.

**Comment by IpswichMapper** [comment]
> This system would also not work on streets where housenumbers
>
> ...?

I meant to say that it would not work for streets that use a housenumber system where it goes up and down, e.g.:

1   2   3    4    5     6
----Ipswich Road------
12  11  10  9  8   7

Where the housenumbers go up by increments of one to the end of the street and then go to the other side, also going up in increments of one.

Also, for interpolation, how could one make that work with Streetcomplete? Maybe a challenge that asks for an address at the edge of each block?

The problem with only doing one side of the street is that you have to walk down the street twice. In other words, unless mapping one side is more than 2x faster than mapping both sides, there is no effeciency gain. 

Currently, like I said, I can do 3km in an hour while mapping both sides. In order to make mapping one side more efficient, I would have to probably be able to map at 7kmph. I don't know about you, but I don't walk that fast normally - I doubt I would do it during streetcomplete. I guess with your system of buttons, it would be possible to map one side of housenumbers while jogging, but I think it is better to get something that can do both sides, to avoid having to jog up and down a street.

**Comment by peternewman** [comment]
> > This system would also not work on streets where housenumbers
> > ...?
> 
> I meant to say that it would not work for streets that use a housenumber system where it goes up and down,

Yes it would @IpswichMapper, that's why I suggested -1/+1 buttons too, for streets that aren't odd and even.

> Also, for interpolation, how could one make that work with Streetcomplete? Maybe a challenge that asks for an address at the edge of each block?

I don't think we should, tagging individual numbers on outlines is by far the most valuable, I think interpolation was only designed as a quick workaround when that's not possible, or when starting out mapping an area. I'd hope no-one would draw an interpolation line with no addresses, so there's no quest for SC to complete about it.
 
> The problem with only doing one side of the street is that you have to walk down the street twice. In other words, unless mapping one side is more than 2x faster than mapping both sides, there is no effeciency gain.
> 
> Currently, like I said, I can do 3km in an hour while mapping both sides. In order to make mapping one side more efficient, I would have to probably be able to map at 7kmph. I don't know about you, but I don't walk that fast normally - I doubt I would do it during streetcomplete. I guess with your system of buttons, it would be possible to map one side of housenumbers while jogging, but I think it is better to get something that can do both sides, to avoid having to jog up and down a street.

Yep, I was suspicious about how practical it would be to do both sides at once, but its possible on a residential road. I suspect for some bigger roads, or with wider verges, you might need to walk it twice. Anyway as I suggested before you just track two sequences then, and have say left and right buttons to switch which sequence you're working off (i.e. which side of the road you're doing addresses for), or it could even alternate each time you submit a quest. Or it could try and be clever and see if you're tagging numbers either side of a road line.

Or perhaps even simpler, just list the last two preceding options with their buttons:

-|Last|house| |One before|last|house|-
--|--|--|--|--|--|--|--
--|3|--|--|--|11|--|--
-2|-1|+1|+2|-2|-1|+1|+2

**Comment by westnordost** [comment]
Ok, I want to add a very important thing here:

#1 priority for the UI design of StreetComplete is easyness and simpleness, not efficiency. Of course we can try to make up a more efficient UI, but only as long as it doesn't negatively effect the most important design goals.

For the +1/-1 idea, maybe make a screenshot/mockup to see how it looks. Question is also if pressing that +1/-1 button should prefill the housenumber edit text and then vanish or if it should stay there so you can repeatedly press +1/-1.

Right now what I am considering is to always expand the keyboard when the quest is opened except for the first time you answer this quest for every session (application start) and make the OK-button in the keyboard act as if you pressed the orange OK button, basically pretty much what @IpswichMapper proposed except that the keyboard is not shown the first time.

**Comment by westnordost** [comment]
A suggestion

![Screenshot_1602585886](https://user-images.githubusercontent.com/4661658/95851094-11ad9500-0d52-11eb-925a-4d631a905dcc.png)

**Comment by smichel17** [comment]
You could pre-fill the previous house number as the "hint" in the text field to make that behavior in-hidden.

**Comment by peternewman** [comment]
> I'm not sure if that was aimed at me, but on that front, shouldn't 123 be the default for house**number**?

As a first step, could we do this? It makes it easier and simpler and is slightly more efficient to boot.

> Ok, I want to add a very important thing here:
> 
> #1 priority for the UI design of StreetComplete is easyness and simpleness, not efficiency. Of course we can try to make up a more efficient UI, but only as long as it doesn't negatively effect the most important design goals.

I'm only suggesting adding a bit more noise to the screen here, similar to the last used height on the building height quest, there is no explanation for either, if you don't press it, you don't need to know what it does.

> For the +1/-1 idea, maybe make a screenshot/mockup to see how it looks. Question is also if pressing that +1/-1 button should prefill the housenumber edit text and then vanish or if it should stay there so you can repeatedly press +1/-1.

You need to be able to do +2/-2 somehow (whether that's a separate button or pressing +1 twice), as most houses (in the UK for example, are odds and evens).

I'm going to make a possibly rather bold statement, I agree with your general design goal of SC, but I think the design goal for the building quests in particular (number, height and to a lesser extent building kind and roof) should actually be efficiency (without making them overly complex).

As quests they head in the direction of monotonous (but valuable), but on a street for example, they need your full attention just to walk at a reasonable speed inputting data, i.e. anyone going out and doing more than one or two of them is going out to survey (so wants speed and efficiency). If you want to get into mapping, many of the other quests are more fun and interesting I'd suggest, you either get properly into it and do lots of house numbers, or you find them tedious and move onto other quests. You can't just do them as a side activity (or there's very little point). In comparison take say the post box, surface, lit, bus stop, traffic light and bench quests, you can do them while walking with other people (I've used the map in SC as my main navigation, perhaps switching to instructions sometimes), or you could go out with your kids "who can find the next bench" etc because they're quick to answer, can keep other people's interest and crucially there aren't too many of them, you can walk a reasonable distance and answer less than ten in some cases.

I would also suggest they are inherently simple because of what they are, type in a number, there's no splitting of ways, or needing to know stuff can be lit 24/7 etc, it's just a visual copy and paste.

> You could pre-fill the previous house number as the "hint" in the text field to make that behavior visible.

I think that could be counter productive either if you're changing sides, or there are ranges of numbers or whatever, you'd have to go clearing it, or get confused, and SC would have to validate that you hadn't re-entered the same value as the last one. @westnordost suggestion of it magically doing that when you press +/- seems a good compromise. I suspect with a lot of these little cheats, and the same with building height, perhaps tied into the achievement system, after you've solved a quest ten times, it wants a "Did you know you can tap the button bottom right to pre-fill the last used buildling height?" type message.

As a bit of a mini poll, as this seems quite key to behaviour, assuming you can read the numbers on both sides from the pavement you're walking down, please like this post as follows:
:rocket: I go down one side (maybe jogging), then do the other side later
:eyes: I do both sides together (i.e. alternating odd and even or increasing and decreasing or whatever)
:tada: I do something else (please comment)

**Comment by peternewman** [comment]
Alternative suggestion (apologies for the budget Gimping):
![image](https://user-images.githubusercontent.com/6216436/95864565-73bfc780-0d5d-11eb-9c00-b897b007d0ae.png)

On your simplification comment, I've dropped the +2 buttons, pressing + twice is still far quicker than typing, and probably easier/less noise on the UI.

The numbers bottom right probably only need to build up to having two past entries (I can't think of a reason for more, and would initially be hidden.

So you answer 101010 and the next time it looks like this screenshot without the 44444, then you answer 44444 and it looks like this screenshot, next time you answer 101011 and it becomes (101011, 44444).

Tapping either historic one populates the box and presents the okay button (like building height), tapping + or - does 44444 +1 or -1 respectively and shows the okay button.

If you enter a range (e.g. 10-14) then -/+ buttons are hidden this time. Next time no history is shown either I suspect. You could do it so pressing + gives you 15 and - gives you 9, or just hide them too.

Again you could do clever fuzzing with 5B->5A/5C or just hide the buttons in that case too.

If it's still deemed too complicated, a title like "History: " (or a better word/phrase) next to the previous ones might help, or a snazzy animation of the number moving to the new position for the first 10 times.

**Comment by westnordost** [comment]
>> I'm not sure if that was aimed at me, but on that front, shouldn't 123 be the default for housenumber?
> As a first step, could we do this? It makes it easier and simpler and is slightly more efficient to boot.

What? If you tap on the housenumber input edit field, it shows the number keyboard. If you press on "abc", it switches to the normal keyboard.

**Comment by westnordost** [comment]
Maybe the question first is what about 

> Right now what I am considering is to always expand the keyboard when the quest is opened except for the first time you answer this quest for every session (application start) and make the OK-button in the keyboard act as if you pressed the orange OK button, basically pretty much what @IpswichMapper proposed except that the keyboard is not shown the first time.

Because both such an UI as you suggest and having the keyboard open immediately makes little sense. If the keyboard is already open, then it is fastest to simply input the number, isn't it?

**Comment by peternewman** [comment]
> > > I'm not sure if that was aimed at me, but on that front, shouldn't 123 be the default for housenumber?
> 
> > As a first step, could we do this? It makes it easier and simpler and is slightly more efficient to boot.
> 
> What? If you tap on the housenumber input edit field, it shows the number keyboard. If you press on "abc", it switches to the normal keyboard.

Ah, in testing at least, I've been tapping on the ABC as the text box already has focus with a flashing cursor. Given there is no keyboard present, making it not have focus would be less confusing, but having the keyboard present would be better with the current form.

> Maybe the question first is what about
> 
> > Right now what I am considering is to always expand the keyboard when the quest is opened except for the first time you answer this quest for every session (application start) and make the OK-button in the keyboard act as if you pressed the orange OK button, basically pretty much what @IpswichMapper proposed except that the keyboard is not shown the first time.
> 
> Because both such an UI as you suggest and having the keyboard open immediately makes little sense. If the keyboard is already open, then it is fastest to simply input the number, isn't it?

No! Maybe I'm particularly rubbish at typing on a phone keyboard (especially while walking along, but I often have to retype building heights too when I type them). Plus if you're in the US with four or five digits, or just a long road in the UK and every house is three digits, that's a lot of typing versus just tapping ++. You've also removed the risk of transposing numbers (no 132) and because you've only got a few buttons I'd imagine your plus is bigger than each keyboard key, so easier to hit.

I think you're right about not being able to see much map too, I can barely see any with the keyboard open, but I'd have over half the screen with my proposed layout.

The keyboard is a good general purpose UI (e.g. house names), but I think having more specialist entry systems for simpler data is the best way to go, why did you bother doing the fancy street name thing otherwise, or the new proposed building heights?

**Comment by westnordost** [comment]
Okay, so we go with the +/- buttons. However, there are a few problems with your last suggestion:
1. when tapping on the button, it pre-fills the field with the **last** (or the one before the last) value, but this should **never** be the final answer. So it is always at minimum three to four taps: first tap on the last number, then tap on plus (two times usually), then on OK. Since the "OK" already appears after the field has been pre-filled, there is even the danger that users will just tap on OK immediately after
2. it makes it a little less obvious (it is already non-obvious) that the +/- button may have the direct effect that it uses the last answered value
3. finally, these two buttons mean that there are now 6 individual controls in the form, 3-4 with numbers on them. This overloads the form a little

So these are the disadvantages I see. Compare it to the suggestion that the keyboard is always open and pressing the "done" button on the keyboard immediately solves the quest: That's usually 3-4 taps as well.

The only advantage your last suggestion has over only displaying the +/- button is that you can efficiently map both sides of the road even when the numbers for the sides are quite far apart. Though, oftentimes, f.e. on larger roads, it is not even well possible to do this because you need to get closer to see the house numbers in the first place. So I don't think this slight advantage is worth the disadvantages described below. Housenumber-mapping both sides on one road is also usually less efficient, more error-prone and more dangerous because one needs to often go zig-zag on the road and it is more difficult to orient yourself (match which house on the map is which house in realitiy). That's my experience at least.

---

When comparing just the +/- vs keyboard is always open + done button solves this quest:

**Pro +/-  / Con keyboard**
- just 2-3 taps
- keyboard is disruptive, shows many buttons
- keyboard hides map, orientation is difficult and might hide important information relevant to the quest

**Pro keyboard / Con +/-**
- just 3-4 taps
- functionality is not obvious, if using a hint, it must be very clear that this is just the hint and it is not pre-filled
- (+) button is very close to the ABC-button / OK-button. Maybe need to move it to the left a bit, but then the input field is not centered anymore which might look a bit unesthetic

**Comment by westnordost** [comment]
Here is another idea for +/- UI.

In green, yet another idea how it could work. Not with tap, but with drag up and down.

![Screenshot_1602600353](https://user-images.githubusercontent.com/4661658/95877215-29961080-0d74-11eb-8030-b2d940768f89.png)

**Comment by peternewman** [comment]
> Okay, so we go with the +/- buttons.

:tada:

> However, there are a few problems with your last suggestion:
> 
>     1. when tapping on the button, it pre-fills the field with the **last** (or the one before the last) value, but this should **never** be the final answer. So it is always at minimum three to four taps: first tap on the last number, then tap on plus (two times usually), then on OK. Since the "OK" already appears after the field has been pre-filled, there is even the danger that users will just tap on OK immediately after

You could only enable okay when the number is not one of the pre-fill ones, so:
Tap last value
Tap +
OK appears
Tap +
Tap OK

>     2. it makes it a little less obvious (it is already non-obvious) that the +/- button may have the direct effect that it uses the last answered value

Agreed, again this depends if you're doing both sides and trying to handle that. I'm just trying to cover the workflows people mentioned.

>     3. finally, these two buttons mean that there are now 6 individual controls in the form, 3-4 with numbers on them. This overloads the form a little

Yeah it didn't feel perfect, I'm hoping it can be refined further by the hive mind.

> So these are the disadvantages I see. Compare it to the suggestion that the keyboard is always open and pressing the "done" button on the keyboard immediately solves the quest: That's usually 3-4 taps as well.
> 
> The only advantage your last suggestion has over only displaying the +/- button is that you can efficiently map both sides of the road even when the numbers for the sides are quite far apart.

In the UK road numbers often seem to end up some what offset, I guess because side roads start shifting it.

> Though, oftentimes, f.e. on larger roads, it is not even well possible to do this because you need to get closer to see the house numbers in the first place.

Yeah on a test walk I found I could read numbers from both sides, I can't remember what I actually did last time, I suspect like you say, one side then the other. Hedges will also break it. Anyway hence my call for votes:
https://github.com/westnordost/StreetComplete/issues/2131#issuecomment-707715784

> So I don't think this slight advantage is worth the disadvantages described below. Housenumber-mapping both sides on one road is also usually less efficient, more error-prone and more dangerous because one needs to often go zig-zag on the road and it is more difficult to orient yourself (match which house on the map is which house in realitiy). That's my experience at least.

Probably agreed.
 
> When comparing just the +/- vs keyboard is always open + done button solves this quest:
> 
> **Pro +/- / Con keyboard**
> 
>     * just 2-3 taps
> 
>     * keyboard is disruptive, shows many buttons
> 
>     * keyboard hides map, orientation is difficult and might hide important information relevant to the quest
> 
> 
> **Pro keyboard / Con +/-**
> 
>     * just 3-4 taps
> 
>     * functionality is not obvious, if using a hint, it must be very clear that this is just the hint and it is not pre-filled
> 
>     * (+) button is very close to the ABC-button / OK-button. Maybe need to move it to the left a bit, but then the input field is not centered anymore which might look a bit unesthetic

Could the ABC button go in the bottom bar opposite Other Answers?

**Comment by peternewman** [comment]
> Here is another idea for +/- UI.
> 
> In green, yet another idea how it could work. Not with tap, but with drag up and down.

Can you tap too or only drag? My experience at work using touchscreens dragging up and down list boxes is far from satisfactory, it's very easy to overshoot. They also don't feel as compatible with data entry while walking along to me.

Also given this form is already at the bottom of the screen, would dragging down for -2 actually work, won't you fall off (or start triggering the Android bar below, or be blocked by that depending on how it all interacts)?

If we're dropping the idea of both sides (or not specifically trying to make it easy), could we look back at one of my previous suggestions, +1 and +2?

Based on yours, move the ABC somewhere; up, down, in that bottom bar, then change the - and + to be -1/+1 and add -2/+2 at the edges. It feels like we're in general agreement that a lot of the time you actually want to add two (so why not give a button specifically for it) and sometimes you want to add 1, and between both of those you can add any other number if needed, so why bother with the drag?
> ![Screenshot_1602585886](https://user-images.githubusercontent.com/4661658/95851094-11ad9500-0d52-11eb-925a-4d631a905dcc.png)

**Comment by westnordost** [comment]
> Could the ABC button go in the bottom bar opposite Other Answers?

No, this breaks expectation of the UI. The bottom bar is for answers only.

**Comment by westnordost** [comment]
I proposed to have both + and - on the right side because it is easier to reach there with the thumb, plus that the edit text is the not centered anymore (on smaller screens) wouldn't look odd. Plus, also thinking about how the form looks like in Czechia for example (there are two input edit texts), maybe 4 +/- buttons wouldn't fit all in the horizontal space available.

**Comment by peternewman** [comment]
> > Could the ABC button go in the bottom bar opposite Other Answers?
> 
> No, this breaks expectation of the UI. The bottom bar is for answers only.

Fair enough.

> I proposed to have both + and - on the right side because it is easier to reach there with the thumb,

You'll have to start at left handed variations next! :smile: 

> plus that the edit text is not centered anymore (on smaller screens) wouldn't look odd. Plus, also thinking about how the form looks like in Czechia for example (there are two input edit texts), maybe 4 +/- buttons wouldn't fit all in the horizontal space available.

Could you stack the two fields on top of each other for the Czech Republic? Although do both their fields increment regularly, I've no idea of their numbering, or is one a block number that only changes every 20 or 50 say?

Your comment elsewhere of hiding ABC sounded good, arguably you could swap -2/-1/+1/+2 for ABC depending on if the keyboard was visible or not.

**Comment by IpswichMapper** [comment]
What is being used? How will you be able to map both sides?

> So I don't think this slight advantage is worth the disadvantages described below. Housenumber-mapping both sides on one road is also usually less efficient, more error-prone and more dangerous because one needs to often go zig-zag on the road and it is more difficult to orient yourself (match which house on the map is which house in realitiy). That's my experience at least.

From my GPX data, I can do around 3km in an hour mapping both sides. In order to be more efficient while mapping one side, I would have to do 6kmph+, which I don't think is possible (without running) with Streetcomplete. 

Furthermore, usually I can quite easily see housenumbers on the other side of the road, even when a road is three lanes wide & a major road.

**Comment by westnordost** [comment]
So this is live now in v25.0-beta1. Just to summarize what I did, and why I chose exactly this UI:

**What I did:**
In Japan:
![Screenshot_20201019-171108_StreetComplete Dev](https://user-images.githubusercontent.com/4661658/96470249-60659e00-122e-11eb-8848-cd0cbb7d9354.jpg)

In the rest of the world:
![Screenshot_20201019-170842_StreetComplete Dev](https://user-images.githubusercontent.com/4661658/96470222-5a6fbd00-122e-11eb-8118-8b4481ae9a3d.jpg)

- the previous input if there is one is shown very slightly in the input field. If the user taps on "+", the input is filled with the next number and the OK button appears. "34a" becomes "35", "10-16" becomes "17" etc. You can repeatedly tap on "+" to count up. The same with tapping "-" but counting down.
- if there was no previous input since application start, the +/- buttons don't do anything
- the ABC-button which is used to switch the keyboard between numbers and any letters is not shown while the keyboard is not open and also only shown for inputs where it is allowed in general to also input letters. (For example for the conscription numbers, it isn't). So it should be clear to users, that this is not the "show keyboard button"
- for Japan, the previous block number is filled in (it's like the "street", in Japan few roads have streets, addressing works by blocks), the field for the housenumber has the same behavior as for the rest of the world
- I changed nothing after all for Czech Republic and Slovakia (with their conscription and orientation numbers)

**Why**
- Czech Republic and Slovakia:  because 1. the +/- buttons wouldn't fit on all screen sizes and 2. you *have to* specify the conscription number anyway always and usually you have to open the keyboard for that anyway. So, there is not really any shortcut that can be taken for the input in those two countries.
- Japan: within one block, the block number never changes, so it is a shortcut to pre-fill it with the last one selected
- the +/- buttons are on the right side so that they are easily reached with the right thumb and the right thumb does not cover up the edit text itself (f.e. if the "-" would be on the left side, while tapping you wouldn't see what's in the input text because your thumb is over it)
- there is no "++" button even though in most cases, the next housenumber from "12" will be "14" (on the same side) etc. because a "double click" will work just as well
- there is no "prefill with selection before the last selection" button because the decision now has been not to encourage zig-zagging along the street but instead to streamline to record all the housenumbers on one side for the reasons stated in https://github.com/westnordost/StreetComplete/issues/2131#issuecomment-707789178
- why the next housenumber of "34a" is "35" and not "34b"? According to taginfo, it is slightly more likely that if there is a house with "34a", the next will be "34b" and not "36". But it is already more likely that the next housenumber of "34b" will be "36" than "34c" and so on. So it is consistent to always count up only the numbers and don't try to be smart with conditionally counting up the letters too.

**Comment by michaelblyons** [comment]
I just wanted to check back and say "Thank you" for the design and programming. I've been using the new interface, and it's really helpful! ❤️

**Comment by westnordost** [comment]
Nice, such feedback is very appreciated :-)

**Comment by thymaro** [comment]
Hey! I came across this while looking for another issue concerning house numbers and thought I'd stop to say this has been implemented really well, at least in Switzerland it's very easy. (I say this, because, looking at the screenshots above, it looks like this is not as easy in every country, so, kudos for that!)
